velox/exec/tests/TableWriteTest.cpp

```diff
@@ -228,40 +291,258 @@ TEST_F(TableWriteTest, constantVectors) {
+  for (int i = 0; i < numBatches; i++) {
+    writeToFile(inputFilePaths[i]->path, vectors[i]);
+  }
+
+  auto outputDirectory = TempDirectoryPath::create();
+  auto plan = PlanBuilder()
+                  .tableScan(rowType)
+                  .tableWrite(
+                      rowType->names(),
+                      std::make_shared<core::InsertTableHandle>(
+                          kHiveConnectorId,
+                          makeHiveInsertTableHandle(
+                              rowType->names(),
+                              rowType->children(),
+                              partitionKeys,
+                              makeLocationHandle(outputDirectory->path))),
+                      CommitStrategy::kNoCommit,
+                      "rows")
+                  .project({"rows"})
+                  .planNode();
+
+  auto task = assertQuery(plan, inputFilePaths, "SELECT count(*) FROM tmp");
+
+  std::set<std::string> partitionDirectories =
+      getLeafSubdirectories(outputDirectory->path);
+
+  // Verify only a single partition directory is created.
+  EXPECT_EQ(partitionDirectories.size(), 1);
+  EXPECT_EQ(
+      *partitionDirectories.begin(),
+      fs::path(outputDirectory->path) / "p0=365");
+
+  // Verify all data is written to the single partition directory.
   assertQuery(
-      PlanBuilder().tableScan(rowType_).planNode(),
+      PlanBuilder().tableScan(rowType).planNode(),
       makeHiveConnectorSplits(outputDirectory),
       "SELECT * FROM tmp");
+
+  // Verify that files written to each partition directory is per driver.
+  EXPECT_EQ(
+      countFilesRecursively(*partitionDirectories.begin()),
+      task->numOutputDrivers());
+}
+
+TEST_F(TableWriteTest, maxPartitions) {
+  std::unordered_map<std::string, std::string> hiveConnectorConfigs{
+      {HiveConfig::kMaxPartitionsPerWriters, "1000"}};
+  int32_t numPartitions = 1'000 + 1;
+
+  auto rowType = ROW({"p0"}, {BIGINT()});
+  std::vector<std::string> partitionKeys = {"p0"};
+
+  auto vector = makeRowVector(
+      rowType->names(),
+      {makeFlatVector<int64_t>(numPartitions, [&](auto row) { return row; })});
+
+  auto outputDirectory = TempDirectoryPath::create();
+  auto plan = PlanBuilder()
+                  .values({vector})
+                  .tableWrite(
+                      rowType->names(),
+                      std::make_shared<core::InsertTableHandle>(
+                          kHiveConnectorId,
+                          makeHiveInsertTableHandle(
+                              rowType->names(),
+                              rowType->children(),
+                              partitionKeys,
+                              makeLocationHandle(outputDirectory->path))),
+                      CommitStrategy::kNoCommit,
+                      "rows")
+                  .project({"rows"})
+                  .planNode();
+
+  CursorParameters params;
+  params.planNode = plan;
+  params.queryCtx = std::make_shared<QueryCtx>(
```

Comment:
Use AssertQueryBuilder

Reply:
This is connector level config (rather than query config) and AssertQueryBuilder right now doesn't support connectorConfig. I'm adding connectorConfig to AssertQueryBuilder easily. But it eventually needs to add a setConnectorConfigOverridesUnsafe() API to QueryCtx.h. Let me know if you think it's not worthwhile and let me revert.

Reply:
> I'm adding connectorConfig to AssertQueryBuilder

Thanks. These changes look good.